File: clojure/day2.clj
(ns aoc-2020.day2
  (:require [clojure.string :as s]))

(defn process-password-entry [entry]
  (let [[policy password]  (s/split entry #":")
        [frequencies char] (s/split policy #" ")
        [min max]          (-> frequencies
                               (s/split #"-")
                               (#(map read-string %)))]
    [[min max] char (s/trim password)]))

(def my-input (-> (slurp "/home/patrick/Documents/projects/1problem-3languages/day2_input.txt")
                  (s/split #"\n")
                  (#(map process-password-entry %))))

;; PART 1
(defn check-password 
  "Takes an password entry which can be explained in the following order:
  character policy, character in a password and the password.
  The character policy indicates required minimum and maximum occurence of the character in the given password.
  Return true if the occurence of the character greater or equal then the minimum value
  and equal or less than maximum value"
  [entry]
  (let [[frequencies char password] entry
        char-count                  (->> password
                                         (#(s/split % #""))
                                         (filter #(= char %))
                                         (count))]
    (and (<= (first frequencies) char-count)
         (>= (second frequencies) char-count))))

(->> my-input
     (map check-password)
     (filter true?)
     (count)) ; => 640

;; PART 2
(defn repeat-exactly-once? [entry]
  (let [[indexes char password] entry
        exist-at-position?      (fn [c s pos] (= c (str (nth s pos))))
        exist-at-indexes        (->> indexes
                                     ;; index of this problem begins with 1
                                     (map dec)
                                     (map #(exist-at-position? char password %)))]
    ;; need to check the result sequence doesn't contain all true values or all false values
    (and (not (every? true? exist-at-indexes))
         (not (every? false? exist-at-indexes)))))

(->> my-input
     (map repeat-exactly-once?)
     (filter true?)
     (count)) ; => 472

File: python/day2.py
with open("/home/patrick/Documents/projects/1problem-3languages/day2_input.txt", "r") as f:
    file_content = f.readlines()


def process_input(item):
    policy, password = item.replace("\n", "").split(":")
    char_policy, character = policy.split(" ")
    char_frequency = list(map(int, char_policy.split("-")))

    output = [char_frequency, character, password.strip()]

    return output


processed_input = [process_input(item) for item in file_content]


# PART 1
def check_password(entry):
    '''Takes an password entry which can be explained in the following order:
       character policy, character in a password and the password.
       The character policy indicates required minimum and maximum occurence of the character in the given password.
       Return true if the occurence of the character greater or equal then the minimum value
       and equal or less than maximum value'''
    if len(entry) == 3:
        min_fre, max_fre = entry[0]
        char = entry[1]
        password = entry[2]

        return (min_fre <= password.count(char)) and (password.count(char) <= max_fre)
    else:
        return false


# Note that filter function always returns a result as an iterator.
# That's why we need to convert the result to a list if we want to use count for it.
correct_password = list(filter(check_password, processed_input))
print(len(correct_password))

# PART 2
def exist_at_position(char, password, index):
    return char == password[index]


def repeat_exactly_once(char, password, indexes):
    repeat = 0
    for index in indexes:
        if exist_at_position(char, password, index):
            repeat += 1

    return repeat == 1


def check_password2(entry):
    policy, character, password = entry

    indexes = [policy[0]-1, policy[1]-1]

    return repeat_exactly_once(character, password, indexes)


correct_password = list(filter(check_password2, processed_input))
print(len(correct_password))
